src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -0,0 +1,258 @@
+        private volatile bool _disposed;
+        private bool _automaticRedirection = true;
+        #endregion
+
+        internal CurlHandler()
+        {
+        }
+
+        #region Properties
+        internal bool AutomaticRedirection
+        {
+            get
+            {
+                return _automaticRedirection;
+            }
+
+            set
+            {
+                CheckDisposedOrStarted();
+                _automaticRedirection = value;
+            }
+        }
+        #endregion
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && !_disposed)
+            {
+                _disposed = true;
+            }
+
+            base.Dispose(disposing);
+        }
+
+        protected internal override Task<HttpResponseMessage> SendAsync(
+            HttpRequestMessage request,
+            CancellationToken cancellationToken)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException("request", SR.net_http_handler_norequest);
+            }
+            if (request.RequestUri.Scheme == UriSchemeHttps)
+            {
+                throw NotImplemented.ByDesignWithMessage("HTTPS stack is not yet implemented");
+            }
+            if (request.Content != null)
+            {
+                throw NotImplemented.ByDesignWithMessage("HTTP requests with a body are not yet supported");
+            }
+
+            CheckDisposed();
+
+            SetOperationStarted();
+
+            TaskCompletionSource<HttpResponseMessage> tcs = new TaskCompletionSource<HttpResponseMessage>();
+
+            // Create RequestState object and save current values of handler settings.
+            RequestState state = new RequestState();
+            state.Tcs = tcs;
+            state.CancellationToken = cancellationToken;
+            state.RequestMessage = request;
+
+            try
+            {
+                Task.Factory.StartNew(
+                    StartRequest,
+                    state,
+                    CancellationToken.None,
+                    TaskCreationOptions.DenyChildAttach,
+                    TaskScheduler.Default);
+            }
+            catch (Exception ex)
+            {
+                throw new HttpRequestException("SendAsync", ex);
+            }
+
+            return tcs.Task;
+        }
+
+        #region Private methods
+
+        private void StartRequest(object obj)
+        {
+            RequestState state = (RequestState)obj;
+            HttpResponseMessage responseMessage = null;
+            Exception savedException = null;
+            SafeCurlHandle requestHandle = null;
+
+            if (state.CancellationToken.IsCancellationRequested)
+            {
+                state.Tcs.TrySetCanceled();
+                return;
+            }
+
+            var cancellationTokenRegistration = state.CancellationToken.Register(() => { state.Tcs.TrySetCanceled(); });
```

Comment:
I believe this will cause a leak of the token which was not there when it was stored in a var and Disposed later. I was thinking you were referring to storing it as a var. What does "closure allocation" mean?